exec/linux/compare/spot/man/randltl.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH RANDLTL "1" "April 2023" "randltl (spot) 2.11.5" "User Commands"
.SH NAME
randltl \- generate random LTL/PSL formulas
.SH SYNOPSIS
.B randltl
[\fI\,OPTION\/\fR...] \fI\,N|PROP\/\fR...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Generate random temporal logic formulas.
.PP
The formulas are built over the atomic propositions named by PROPS...
or, if N is a nonnegative number, using N arbitrary names.
.SS "Type of formula to generate:"
.TP
\fB\-B\fR, \fB\-\-boolean\fR
generate Boolean formulas
.TP
\fB\-L\fR, \fB\-\-ltl\fR
generate LTL formulas (default)
.TP
\fB\-P\fR, \fB\-\-psl\fR
generate PSL formulas
.TP
\fB\-S\fR, \fB\-\-sere\fR
generate SERE
.SS "Generation:"
.TP
\fB\-\-allow\-dups\fR
allow duplicate formulas to be output
.TP
\fB\-n\fR, \fB\-\-formulas\fR=\fI\,INT\/\fR
number of formulas to output (1)
use a negative value for unbounded generation
.TP
\fB\-r\fR, \fB\-\-simplify\fR[=\fI\,LEVEL\/\fR]
simplify formulas according to LEVEL (see below);
LEVEL is set to 3 if omitted
.TP
\fB\-\-seed\fR=\fI\,INT\/\fR
seed for the random number generator (0)
.TP
\fB\-\-tree\-size\fR=\fI\,RANGE\/\fR
tree size of the formulas generated, before
mandatory trivial simplifications (15)
.TP
\fB\-\-weak\-fairness\fR
append some weak\-fairness conditions
.PP
RANGE may have one of the following forms: 'INT', 'INT..INT', or '..INT'.
In the latter case, the missing number is assumed to be 1.
.PP
The simplification LEVEL may be set as follows.
.TP
0
No rewriting
.TP
1
basic rewritings and eventual/universal rules
.TP
2
additional syntactic implication rules
.TP
3
better implications using containment
.SS "Adjusting probabilities:"
.TP
\fB\-\-boolean\-priorities\fR=\fI\,STRING\/\fR
set priorities for Boolean formulas
.TP
\fB\-\-dump\-priorities\fR
show current priorities, do not generate any
formula
.TP
\fB\-\-ltl\-priorities\fR=\fI\,STRING\/\fR
set priorities for LTL formulas
.TP
\fB\-\-sere\-priorities\fR=\fI\,STRING\/\fR
set priorities for SERE formulas
.PP
STRING should be a comma\-separated list of assignments, assigning integer
priorities to the tokens listed by \fB\-\-dump\-priorities\fR.
.SS "Output options:"
.TP
\fB\-0\fR, \fB\-\-zero\-terminated\-output\fR
separate output formulas with \e0 instead of \en
(for use with xargs \fB\-0\fR)
.TP
\fB\-8\fR, \fB\-\-utf8\fR
output using UTF\-8 characters
.TP
\fB\-\-format\fR=\fI\,FORMAT\/\fR, \fB\-\-stats\fR=\fI\,FORMAT\/\fR
specify how each line should be output (default:
"%f")
.TP
\fB\-l\fR, \fB\-\-lbt\fR
output in LBT's syntax
.TP
\fB\-\-latex\fR
output using LaTeX macros
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,FORMAT\/\fR
send output to a file named FORMAT instead of
standard output.  The first formula sent to a file
truncates it unless FORMAT starts with '>>'.
.TP
\fB\-p\fR, \fB\-\-full\-parentheses\fR
output fully\-parenthesized formulas
.TP
\fB\-s\fR, \fB\-\-spin\fR
output in Spin's syntax
.TP
\fB\-\-spot\fR
output in Spot's syntax (default)
.TP
\fB\-\-wring\fR
output in Wring's syntax
.PP
The FORMAT string passed to \fB\-\-format\fR may use the following interpreted
sequences:
.TP
%%
a single %
.TP
%b
the Boolean\-length of the formula (i.e., all
Boolean subformulas count as 1)
.TP
%f
the formula (in the selected syntax)
.TP
%h, %[vw]h
the class of the formula is the Manna\-Pnueli
hierarchy ([v] replaces abbreviations by class
names, [w] for all compatible classes)
.TP
%L
the (serial) number of the formula
.TP
%[OP]n
the nesting depth of operator OP.  OP should be a
single letter denoting the operator to count, or
multiple letters to fuse several operators during
depth evaluation.  Add '~' to rewrite the formula
in negative normal form before counting.
.TP
%s
the length (or size) of the formula
.TP
%x, %[LETTERS]X, %[LETTERS]x
number of atomic propositions used in the
.TP
formula;
add LETTERS to list atomic propositions
.TP
with (n) no quoting, (s) occasional double\-quotes
with C\-style escape, (d) double\-quotes with
C\-style escape, (c) double\-quotes with CSV\-style
escape, (p) between parentheses, any extra
non\-alphanumeric character will be used to
separate propositions
.SS "Miscellaneous options:"
.TP
\fB\-\-help\fR
print this help
.TP
\fB\-\-version\fR
print program version
.PP
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.SH BIBLIOGRAPHY
If you would like to give a reference to this tool in an article,
we suggest you cite the following paper:
.TP
\(bu
Alexandre Duret-Lutz: Manipulating LTL formulas using Spot 1.0.
Proceedings of ATVA'13.  LNCS 8172.
.SH EXAMPLES
The following generates 10 random LTL formulas over the propositions a, b,
and c, with the default tree\-size, and all available operators.
.IP
\f(CW% randltl -n10 a b c\fR
.PP
If you do not mind about the name of the atomic propositions, just give
a number instead:
.IP
\f(CW% randltl -n10 3\fR
.PP
You can disable or favor certain operators by changing their priority.
The following disables xor, implies, and equiv, and multiply the probability
of X to occur by 10.
.IP
\f(CW% randltl --ltl-priorities='xor=0, implies=0, equiv=0, X=10' -n10 a b c\fR
.SH "REPORTING BUGS"
Report bugs to <spot@lrde.epita.fr>.
.SH COPYRIGHT
Copyright \(co 2023 Laboratoire de Recherche de l'Epita (LRE)
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR genaut (1),
.BR genltl (1),
.BR ltlfilt (1),
.BR randaut (1)


exec/linux/compare/spot-new/randltl.cc
// -*- coding: utf-8 -*-
// Copyright (C) 2012-2016, 2018-2019, 2022, 2023 Laboratoire de
// Recherche et Développement de l'Epita (LRDE).
//
// This file is part of Spot, a model checking library.
//
// Spot is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// Spot is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
// License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "common_sys.hh"

#include <iostream>
#include <fstream>
#include <argp.h>
#include <cstdlib>
#include <iterator>
#include "error.h"

#include "common_setup.hh"
#include "common_output.hh"
#include "common_range.hh"
#include "common_r.hh"
#include "common_conv.hh"
#include "common_cout.hh"

#include <sstream>
#include <spot/tl/defaultenv.hh>
#include <spot/tl/randomltl.hh>
#include <spot/tl/simplify.hh>
#include <spot/misc/random.hh>
#include <spot/misc/optionmap.hh>

static const char argp_program_doc[] = "\
Generate random temporal logic formulas.\n\n\
The formulas are built over the atomic propositions named by PROPS...\n\
or, if N is a nonnegative number, using N arbitrary names.\v\
Examples:\n\
\n\
The following generates 10 random LTL formulas over the propositions a, b,\n\
and c, with the default tree-size, and all available operators.\n\
  % randltl -n10 a b c\n\
\n\
If you do not mind about the name of the atomic propositions, just give\n\
a number instead:\n\
  % randltl -n10 3\n\
\n\
You can disable or favor certain operators by changing their priority.\n\
The following disables xor, implies, and equiv, and multiply the probability\n\
of X to occur by 10.\n\
  % randltl --ltl-priorities='xor=0, implies=0, equiv=0, X=10' -n10 a b c\n\
";

enum {
  OPT_BOOLEAN_PRIORITIES = 1,
  OPT_DUMP_PRIORITIES,
  OPT_DUPS,
  OPT_LTL_PRIORITIES,
  OPT_SEED,
  OPT_SERE_PRIORITIES,
  OPT_TREE_SIZE,
  OPT_WF,
};

static const argp_option options[] =
  {
    // Keep this alphabetically sorted (expect for aliases).
    /**************************************************/
    { nullptr, 0, nullptr, 0, "Type of formula to generate:", 1 },
    { "boolean", 'B', nullptr, 0, "generate Boolean formulas", 0 },
    { "ltl", 'L', nullptr, 0, "generate LTL formulas (default)", 0 },
    { "sere", 'S', nullptr, 0, "generate SERE", 0 },
    { "psl", 'P', nullptr, 0, "generate PSL formulas", 0 },
    /**************************************************/
    { nullptr, 0, nullptr, 0, "Generation:", 2 },
    { "weak-fairness", OPT_WF, nullptr, 0,
      "append some weak-fairness conditions", 0 },
    { "formulas", 'n', "INT", 0, "number of formulas to output (1)\n"\
      "use a negative value for unbounded generation", 0 },
    { "seed", OPT_SEED, "INT", 0,
      "seed for the random number generator (0)", 0 },
    { "tree-size", OPT_TREE_SIZE, "RANGE", 0,
      "tree size of the formulas generated, before mandatory "\
      "trivial simplifications (15)", 0 },
    { "allow-dups", OPT_DUPS, nullptr, 0,
      "allow duplicate formulas to be output", 0 },
    DECLARE_OPT_R,
    RANGE_DOC,
    LEVEL_DOC(3),
    /**************************************************/
    { nullptr, 0, nullptr, 0, "Adjusting probabilities:", 4 },
    { "dump-priorities", OPT_DUMP_PRIORITIES, nullptr, 0,
      "show current priorities, do not generate any formula", 0 },
    { "ltl-priorities", OPT_LTL_PRIORITIES, "STRING", 0,
      "set priorities for LTL formulas", 0 },
    { "sere-priorities", OPT_SERE_PRIORITIES, "STRING", 0,
      "set priorities for SERE formulas", 0 },
    { "boolean-priorities", OPT_BOOLEAN_PRIORITIES, "STRING", 0,
      "set priorities for Boolean formulas", 0 },
    { nullptr, 0, nullptr, 0, "STRING should be a comma-separated list of "
      "assignments, assigning integer priorities to the tokens "
      "listed by --dump-priorities.", 0 },
    /**************************************************/
    { nullptr, 0, nullptr, 0, "Output options:", -20 },
    { nullptr, 0, nullptr, 0, "The FORMAT string passed to --format may use "
      "the following interpreted sequences:", -19 },
    { "%f", 0, nullptr, OPTION_DOC | OPTION_NO_USAGE,
      "the formula (in the selected syntax)", 0 },
    { "%L", 0, nullptr, OPTION_DOC | OPTION_NO_USAGE,
      "the (serial) number of the formula", 0 },
    { "%%", 0, nullptr, OPTION_DOC | OPTION_NO_USAGE,
      "a single %", 0 },
    COMMON_LTL_OUTPUT_SPECS,
    /**************************************************/
    { nullptr, 0, nullptr, 0, "Miscellaneous options:", -1 },
    { nullptr, 0, nullptr, 0, nullptr, 0 }
  };


const struct argp_child children[] =
  {
    { &output_argp, 0, nullptr, 0 },
    { &misc_argp, 0, nullptr, 0 },
    { nullptr, 0, nullptr, 0 }
  };

// We want all these variables to be destroyed when we exit main, to
// make sure it happens before all other global variables (like the
// atomic propositions maps) are destroyed.  Otherwise we risk
// accessing deleted stuff.
static struct opt_t
{
  spot::atomic_prop_set aprops;
}* opt;

static spot::randltlgenerator::output_type output =
  spot::randltlgenerator::LTL;
static char* opt_pL = nullptr;
static char* opt_pS = nullptr;
static char* opt_pB = nullptr;
static bool opt_dump_priorities = false;
static int opt_formulas = 1;
static int opt_seed = 0;
static range opt_tree_size = { 15, 15 };
static bool opt_unique = true;
static bool opt_wf = false;
static bool ap_count_given = false;

static int
parse_opt(int key, char* arg, struct argp_state* as)
{
  // Called from C code, so should not raise any exception.
  BEGIN_EXCEPTION_PROTECT;
  // This switch is alphabetically-ordered.
  switch (key)
    {
    case 'B':
      output = spot::randltlgenerator::Bool;
      break;
    case 'L':
      output = spot::randltlgenerator::LTL;
      break;
    case 'n':
      opt_formulas = to_int(arg, "-n/--formulas");
      break;
    case 'P':
      output = spot::randltlgenerator::PSL;
      break;
    case OPT_R:
      parse_r(arg);
      break;
    case 'S':
      output = spot::randltlgenerator::SERE;
      output_ratexp = true;
      break;
    case OPT_BOOLEAN_PRIORITIES:
      opt_pB = arg;
      break;
    case OPT_DUPS:
      opt_unique = false;
      break;
    case OPT_LTL_PRIORITIES:
      opt_pL = arg;
      break;
    case OPT_DUMP_PRIORITIES:
      opt_dump_priorities = true;
      break;
    case OPT_SERE_PRIORITIES:
      opt_pS = arg;
      break;
    case OPT_SEED:
      opt_seed = to_int(arg, "--seed");
      break;
    case OPT_TREE_SIZE:
      opt_tree_size = parse_range(arg);
      if (opt_tree_size.min > opt_tree_size.max)
        std::swap(opt_tree_size.min, opt_tree_size.max);
      break;
    case OPT_WF:
      opt_wf = true;
      break;
    case ARGP_KEY_ARG:
      // If this is the unique non-option argument, it can
      // be a number of atomic propositions to build.
      //
      // argp reorganizes argv[] so that options always come before
      // non-options.  So if as->argc == as->next we know this is the
      // last non-option argument, and if aprops.empty() we know this
      // is the also the first one.
      if (opt->aprops.empty() && as->argc == as->next)
        {
          char* endptr;
          int res = strtol(arg, &endptr, 10);
          if (!*endptr && res >= 0) // arg is a number
            {
              ap_count_given = true;
              opt->aprops = spot::create_atomic_prop_set(res);
              break;
            }
        }
      opt->aprops.insert(spot::default_environment::instance().require(arg));
      break;
    default:
      return ARGP_ERR_UNKNOWN;
    }
  END_EXCEPTION_PROTECT;
  return 0;
}

int
main(int argc, char** argv)
{
  return protected_main(argv, [&] {
      const argp ap = { options, parse_opt, "N|PROP...", argp_program_doc,
                        children, nullptr, nullptr };

      // This will ensure that all objects stored in this struct are
      // destroyed before global variables.
      opt_t o;
      opt = &o;

      if (int err = argp_parse(&ap, argc, argv, ARGP_NO_HELP, nullptr, nullptr))
        exit(err);

      // running 'randltl 0' is one way to generate formulas using no
      // atomic propositions so do not complain in that case.
      if (opt->aprops.empty() && !ap_count_given)
        error(2, 0, "No atomic proposition supplied?  "
              "Run '%s --help' for usage.", program_name);

      spot::srand(opt_seed);

      spot::randltlgenerator rg
        (opt->aprops,
         [&] (){
          spot::option_map opts;
          opts.set("output", output);
          opts.set("tree_size_min", opt_tree_size.min);
          opts.set("tree_size_max", opt_tree_size.max);
          opts.set("wf", opt_wf);
          opts.set("seed", opt_seed);
          opts.set("simplification_level", simplification_level);
          opts.set("unique", opt_unique);
          return opts;
        }(), opt_pL, opt_pS, opt_pB);

      if (opt_dump_priorities)
        {
          switch (output)
            {
            case spot::randltlgenerator::LTL:
              std::cout <<
                "Use --ltl-priorities to set the following LTL priorities:\n";
              rg.dump_ltl_priorities(std::cout);
              break;
            case spot::randltlgenerator::Bool:
              std::cout <<
                "Use --boolean-priorities to set the following Boolean "
                "formula priorities:\n";
              rg.dump_bool_priorities(std::cout);
              break;
            case spot::randltlgenerator::PSL:
              std::cout <<
                "Use --ltl-priorities to set the following LTL priorities:\n";
              rg.dump_psl_priorities(std::cout);
              SPOT_FALLTHROUGH;
            case spot::randltlgenerator::SERE:
              std::cout <<
                "Use --sere-priorities to set the following SERE priorities:\n";
              rg.dump_sere_priorities(std::cout);
              std::cout <<
                "Use --boolean-priorities to set the following Boolean "
                "formula priorities:\n";
              rg.dump_sere_bool_priorities(std::cout);
              break;
            }
          exit(0);
        }

      while (opt_formulas < 0 || opt_formulas--)
        {
          static int count = 0;
          spot::formula f = rg.next();
          if (!f)
            {
              error(2, 0, "failed to generate a new unique formula after %d " \
                    "trials", spot::randltlgenerator::MAX_TRIALS);
            }
          else
            {
              output_formula_checked(f, nullptr, nullptr, ++count);
            }
        };
      flush_cout();
      return 0;
    });
}
